### PR TITLE
SPOI-11387 Modifying operator code to incorporate listener tests

### PR DESCRIPTION
**Pull request by KapoorHitesh**
@DT-Priyanka please review

### CODE REVIEW

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java [hunk 1]
```diff
@@ -43,22 +44,74 @@
...
+  public final transient DefaultOutputPort<Map<Rule, Integer>> ruleCountOutput = new DefaultOutputPort<>();
+  public final transient DefaultOutputPort<Map<Object, List<Rule>>> factAndFiredRulesOutput = new DefaultOutputPort<>();
   private transient StatelessKieSession kieSession;
   private String rulesDir;
   private boolean loadSpringSession = false;
   private String sessionName;
   private Map<Object, List<Rule>> factsAndFiredRules;
   private Map<Rule, Integer> ruleCount;
+  private List<Object> facts;
+  private List<Object> factsFromRules;
```
Comment (DT-Priyanka) [inline drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: can we call it as rulesGeneratedFacts or factsGeneratedByRules

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java [hunk 2]
```diff
@@ -43,22 +44,74 @@
...
   };
 
   @Override
+  public void endWindow()
+  {
+    super.endWindow();
+    if (facts.size() > 0) {
+      executeRules();
+    }
+    factAndFiredRulesOutput.emit(factsAndFiredRules);
```
Comment (DT-Priyanka) [inline drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: do we want to send it one by one? same for ruleCount?

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java [hunk 3]
```diff
@@ -43,22 +44,74 @@
...
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    super.beginWindow(windowId);
+    factsAndFiredRules.clear();
+    ruleCount.clear();
+  }
+
+  public int getBatchSize()
```
Comment (DT-Priyanka) [inline drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: put javadoc

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java [hunk 4]
```diff
@@ -75,6 +128,8 @@ public void activate(OperatorContext context)
       KieContainer kieContainer = initializeKieContainerFromRulesDir();
```
Comment (DT-Priyanka) [inline drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java:140]: we should move this to activate.

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java [hunk 5]
```diff
@@ -256,19 +318,24 @@ public void afterRuleFlowGroupDeactivated(RuleFlowGroupDeactivatedEvent ruleFlow
...
+     * When the getRule() will not be null in that case the fact is added via a rule.
      *
      * @param objectInsertedEvent fact inserted in the session
      */
     @Override
     public void objectInserted(ObjectInsertedEvent objectInsertedEvent)
     {
-      factsAndFiredRules.put(objectInsertedEvent.getObject(), new ArrayList<Rule>());
+      Object newFact = objectInsertedEvent.getObject();
+      factsAndFiredRules.put(newFact, new ArrayList<Rule>());
```
Comment (DT-Priyanka) [inline drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: should we put  new fact in this map?? we need to think about it

File: drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java [hunk 6]
```diff
@@ -15,11 +19,18 @@
...
 import com.datatorrent.lib.helper.OperatorContextTestHelper;
+import com.datatorrent.lib.testbench.CollectorTestSink;
 
 public class StatelessDroolsOperatorTest
 {
-  public static class TestMeta extends TestWatcher
+  CollectorTestSink<Object> factsPort = new CollectorTestSink<Object>();
+  CollectorTestSink<Object> factsAndFiredRulesPort = new CollectorTestSink<>();
+  CollectorTestSink<Object> ruleCountPort = new CollectorTestSink<>();
+  StatelessDroolsOperator droolsOperator = new StatelessDroolsOperator();
```
Comment (DT-Priyanka) [inline drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java]: we already have instance of this class, check "underTest"

File: drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java [hunk 7]
```diff
@@ -46,12 +57,19 @@ protected void starting(Description description)
       context = new OperatorContextTestHelper.TestIdOperatorContext(0, attributes);
       underTest = new StatelessDroolsOperator();
       underTest.setRulesDir(rulesDirectory);
+      underTest.setup(null);
+      underTest.factsOutput.setSink(factsPort);
```
Comment (DT-Priyanka) [inline drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java]: you can pass test context to setup.

File: drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java [hunk 8]
```diff
@@ -77,6 +95,8 @@ public void testActivation()
   {
     testMeta.underTest.activate(testMeta.context);
     StatelessKieSession session = testMeta.underTest.getKieSession();
+    Product p = new Product(1,"abc",2);
+    session.execute(p);
```
Comment (DT-Priyanka) [inline drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java]: session.execute in "creatRulesFiles"?? looks out of place.

File: drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java [hunk 9]
```diff
@@ -98,4 +118,119 @@ public void testSpringActivation()
...
+    List<Object> factsList = getListForRuleMatch();
+    testMeta.underTest.beginWindow(0);
+    for (Object fact : factsList) {
+      testMeta.underTest.factsInput.process(fact);
+    }
+    Assert.assertEquals(0,factsPort.collectedTuples.size());
+    testMeta.underTest.endWindow();
+    Assert.assertEquals(1,ruleCountPort.collectedTuples.size());
+    Assert.assertEquals(1,factsAndFiredRulesPort.collectedTuples.size());
+    //It is expected to get 8 tuples as platinum is added via rules
```
Comment (DT-Priyanka) [inline drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java:183]: Inserted fact could be different test. It's a good idea to write a test with one specific test purpose. It's okay to have multiple small tests. When something goes wrong it's easy to nail down if we have separate tests.

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 2 above)
Comment (KapoorHitesh) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: I hope you mean the entries of the map. It doesn't makes sense to do it one by one. The ordering of the tuples is anyways not guranteed

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 5 above)
Comment (KapoorHitesh) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: We do need to put it in this map.

File: drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java (hunk 6 above)
Comment (KapoorHitesh) [reply drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java]: yes will remove it

File: drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java (hunk 8 above)
Comment (KapoorHitesh) [reply drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java]: yes will remove it

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 1 above)
Comment (KapoorHitesh) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: factsFromRules serves the purpose names suggested by you are too big .

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 5 above)
Comment (KapoorHitesh) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: For every fact getting emitted on the factsOutput port need to have a corresponding entry here.

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 4 above)
Comment (KapoorHitesh) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java:140]: Comment looks out of place can.

File: drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java (hunk 9 above)
Comment (KapoorHitesh) [reply drools/src/test/java/com/datatorrent/drools/operator/StatelessDroolsOperatorTest.java:183]: In testing a function/module, it's OK to have multiple related asserts. Have followed the same approach while writing operators in Malhar.

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 2 above)
Comment (DT-Priyanka) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: I was thinking from downsteam operator's perspective. Will it be better if we send it separate for downsteam operators?

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java [hunk 10]
```diff
@@ -43,22 +44,78 @@
...
+  private static final int DEFAULT_BATCH_SIZE = 1000;
+  public final transient DefaultOutputPort<Object> factsOutput = new DefaultOutputPort<>();
+  public final transient DefaultOutputPort<Map<Rule, Integer>> ruleCountOutput = new DefaultOutputPort<>();
+  public final transient DefaultOutputPort<Map<Object, List<Rule>>> factAndFiredRulesOutput = new DefaultOutputPort<>();
   private transient StatelessKieSession kieSession;
   private String rulesDir;
   private boolean loadSpringSession = false;
   private String sessionName;
   private Map<Object, List<Rule>> factsAndFiredRules;
   private Map<Rule, Integer> ruleCount;
```
Comment (bhupeshchawda) [inline drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: can `ruleCount` and `factsAndFiredRules` be transient?

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java [hunk 11]
```diff
@@ -43,22 +44,78 @@
...
+      executeRules();
+    }
+    factAndFiredRulesOutput.emit(factsAndFiredRules);
+    ruleCountOutput.emit(ruleCount);
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    super.beginWindow(windowId);
```
Comment (bhupeshchawda) [inline drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: no need..

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java [hunk 12]
```diff
@@ -43,22 +44,78 @@
...
+  public final transient DefaultOutputPort<Map<Rule, Integer>> ruleCountOutput = new DefaultOutputPort<>();
+  public final transient DefaultOutputPort<Map<Object, List<Rule>>> factAndFiredRulesOutput = new DefaultOutputPort<>();
   private transient StatelessKieSession kieSession;
   private String rulesDir;
   private boolean loadSpringSession = false;
   private String sessionName;
   private Map<Object, List<Rule>> factsAndFiredRules;
   private Map<Rule, Integer> ruleCount;
+  private List<Object> facts;
+  private List<Object> factsFromRules;
```
Comment (bhupeshchawda) [inline drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: Also, `facts` and `factsFromRules` can be transient..?

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java [hunk 13]
```diff
@@ -43,22 +44,78 @@
...
+      if (facts.size() >= batchSize) {
+        executeRules();
+      }
     }
   };
 
   @Override
+  public void endWindow()
+  {
+    super.endWindow();
```
Comment (bhupeshchawda) [inline drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: no need..

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java [hunk 14]
```diff
@@ -256,13 +322,18 @@ public void afterRuleFlowGroupDeactivated(RuleFlowGroupDeactivatedEvent ruleFlow
...
+     * When the getRule() will not be null in that case the fact is added via a rule.
      *
      * @param objectInsertedEvent fact inserted in the session
      */
     @Override
     public void objectInserted(ObjectInsertedEvent objectInsertedEvent)
     {
-      factsAndFiredRules.put(objectInsertedEvent.getObject(), new ArrayList<Rule>());
+      Object newFact = objectInsertedEvent.getObject();
+      factsAndFiredRules.put(newFact, new ArrayList<Rule>());
```
Comment (bhupeshchawda) [inline drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: Can you get this new fact more than once? In other words, can the same fact be generated multiple times in the course of time. If, so can you also check if the fact is already inserted in the `factsAndFiredRules` map?

File: drools/pom.xml [hunk 15]
```diff
@@ -55,6 +55,12 @@
       <scope>test</scope>
     </dependency>
     <dependency>
+      <groupId>org.apache.apex</groupId>
+      <artifactId>malhar-library</artifactId>
```
Comment (bhupeshchawda) [inline drools/pom.xml:59]: where is this used?

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java [hunk 16]
```diff
@@ -43,22 +44,78 @@
 public class StatelessDroolsOperator extends BaseOperator implements ActivationListener<Context.OperatorContext>
```
Comment (bhupeshchawda) [inline drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java:60]: Have we finalized on the license for this operator. If so, please add appropriate license header.. for all files..

**Review by bhupeshchawda — COMMENTED** [review]
Also, please add javadocs to classes, getters/setters and other public methods..

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 14 above)
Comment (KapoorHitesh) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: There is no restriction to add same fact multiple times, will change the code to incorporate your suggestion.

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 2 above)
Comment (KapoorHitesh) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: @bhupeshchawda could you give your opinion on it?

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 10 above)
Comment (KapoorHitesh) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: done

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 12 above)
Comment (KapoorHitesh) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: done

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 13 above)
Comment (KapoorHitesh) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: done

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 11 above)
Comment (KapoorHitesh) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: done

File: drools/pom.xml (hunk 15 above)
Comment (KapoorHitesh) [reply drools/pom.xml:59]: For StatelessDroolsOperatorTest in using CollectorTestSink

File: drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java (hunk 2 above)
Comment (bhupeshchawda) [reply drools/src/main/java/com/datatorrent/drools/operator/StatelessDroolsOperator.java]: It does makes sense to have a single entry at a time and we would have to emit a Map.Entry<>. But then it is also okay for the downstream operator to process a Map. 
I think we can hold on with Maps until we know what exactly would this data be used for.

Also, do mark the class as `@Evolving`..

**Pull request merged**